cirq/google/op_deserializer.py

```diff
@@ -85,25 +88,38 @@ def __init__(
                 no number of qubits is passed to the constructor.
             op_wrapper: An optional Callable to modify the resulting
                 GateOperation, for instance, to add tags
+            auto_deserialize_tokens: Whether to convert tokens to
+                CalibrationTags. Defaults to True.
         """
         self.serialized_gate_id = serialized_gate_id
         self.gate_constructor = gate_constructor
         self.args = args
         self.num_qubits_param = num_qubits_param
         self.op_wrapper = op_wrapper
+        self.auto_deserialize_tokens = auto_deserialize_tokens
 
     def from_proto(self,
                    proto: v2.program_pb2.Operation,
                    *,
-                   arg_function_language: str = '') -> 'cirq.Operation':
+                   arg_function_language: str = '',
+                   constants: List[v2.program_pb2.Constant] = None
+                  ) -> 'cirq.Operation':
         """Turns a cirq.google.api.v2.Operation proto into a GateOperation."""
         qubits = [v2.qubit_from_proto_id(q.id) for q in proto.qubits]
         args = self._args_from_proto(
             proto, arg_function_language=arg_function_language)
         if self.num_qubits_param is not None:
             args[self.num_qubits_param] = len(qubits)
         gate = self.gate_constructor(**args)
-        return self.op_wrapper(gate.on(*qubits), proto)
+        op = self.op_wrapper(gate.on(*qubits), proto)
+        if self.auto_deserialize_tokens:
+            if proto.HasField('constant_index') and constants:
```

Comment:
I would suggest checking which case is populated with `WhichOneof`:
```python
which = proto.WhichOneof('token')
if which == 'token_constant_index':
    ...
elif which == 'token_value':
    ...
```
Also, if a constant index is specified but `constants` is empty, we should probably raise an error.

Reply:
Done.